sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala

```diff
@@ -36,14 +36,14 @@ import org.apache.spark.util.Utils
 case class CollectLimitExec(limit: Int, child: SparkPlan) extends UnaryExecNode {
   override def output: Seq[Attribute] = child.output
   override def outputPartitioning: Partitioning = SinglePartition
-  override def executeCollect(): Array[InternalRow] = child.executeTake(limit)
-  private val serializer: Serializer = new UnsafeRowSerializer(child.output.size)
+  override def requiredChildDistribution: List[Distribution] = AllTuples :: Nil
+  override def executeCollect(): Array[InternalRow] = child match {
+    case e: Exchange => e.child.executeTake(limit)
+    case _ => child.executeTake(limit)
+  }
+
   protected override def doExecute(): RDD[InternalRow] = {
-    val locallyLimited = child.execute().mapPartitionsInternal(_.take(limit))
```

Comment:
We are removing an optimization here right? We can greatly reduce the number of shuffled records by applying the limit before anything gets shuffled.